nodist/clang_check_attributes.py

```diff
@@ -0,0 +1,325 @@
+                tu_call = find_usr(
+                    str(call.location.file).replace(".h", ".cpp"),
+                    call.get_usr(), cfg)
+                if tu_call is not None:
+                    call = tu_call
+
+        if call is not None:
+            if fully_qualified_pretty(call) not in PRINTED:
+                PRINTED.append(fully_qualified_pretty(call))
+
+            if not check_attributes(call, search_attributes):
+                return False
+            if call.kind in [
+                    CursorKind.CXX_METHOD, CursorKind.CONSTRUCTOR,
+                    CursorKind.FUNCTION_DECL
+            ]:
+                return find_funcs(call, cfg, search_attributes)
+    return True
+
+
+def print_funcs(node, cfg, search_attributes, level=1):
+    funcs = get_call_expr(node)
+    for func in funcs:
+        if not func.referenced:
+            continue
+        if is_excluded(func.referenced, cfg):
+            continue
+        call = func.referenced
+        # Don't recurse into already printed nodes
+        # it some cases in increases printing nodes by a lot
+        if fully_qualified_pretty(call) in PRINTED:
+            continue
+        if str(call.location.file).endswith(".h"):
+            if os.path.exists(str(call.location.file).replace(".h", ".cpp")):
+                tu_call = find_usr(
+                    str(call.location.file).replace(".h", ".cpp"),
+                    call.get_usr(), cfg)
+                if tu_call is not None:
+                    call = tu_call
+
+        if call is not None:
+            print_node(call, level)
+            if call.kind in [
+                    CursorKind.CXX_METHOD, CursorKind.CONSTRUCTOR,
+                    CursorKind.FUNCTION_DECL
+            ]:
+                print_funcs(call, cfg, search_attributes, level + 1)
+
+
+def get_info(node, cfg):
+    if node.kind == CursorKind.CXX_METHOD:
+        annotations = get_annotations(node)
+        if "MT_SAFE" in annotations:
+            if not find_funcs(node, cfg, ["MT_SAFE", "PURE"]):
+                UNSAFE_CALLS.append(node)
+        if "PURE" in annotations:
+            if not find_funcs(node, cfg, ["PURE"]):
+                UNSAFE_CALLS.append(node)
+        if "MT_START" in annotations:
+            if not find_funcs(node, cfg, ["MT_SAFE", "PURE"]):
+                UNSAFE_CALLS.append(node)
+    for cursor in node.get_children():
+        get_info(cursor, cfg)
+
+
+def get_files(filename):
+    files = list(glob.glob(f"{filename}/*.cpp"))
+    # Dont process test files
+    files = [f for f in files if "_test.cpp" not in f]
+    # Dont process Verilator coverage files
+    files = [f for f in files if "Vlc" not in f]
+    # We want to use V3Const after astgen
+    files = [f for f in files if "V3Const.cpp" not in f]
+    # Skip SystemC files
+    files = [f for f in files if "_sc" not in f]
+    # Skip parser files
+    files = [f for f in files if ".yy.cpp" not in f]
```

Comment:
Perhaps cleaner to have a list of regexps to ignore at top of program, then iterate that here?